[Bug] [SQL] SQL Task Node Custom Parameters run failed

### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue) and found no similar issues.


### What happened

when i using Custom Parameters of SQL Task Node with OUT type, sql query result is more than one row, in downstream sql using the parameters failed, if upstream sql result is one row, it work fine.
failed work flow:
![image](https://user-images.githubusercontent.com/25889738/172971291-7840b455-7325-4c98-805b-a6b5f8b9ad75.png)

work flow success
![image](https://user-images.githubusercontent.com/25889738/172971336-5ac1bc60-2fb7-4ef5-b996-cbb03f1653d8.png)


### What you expected to happen

SQL Task Custom parameters work fine

### How to reproduce

add two sql task node, in first node query select and add custom parameter OUT type, in donwstream sql using the parameters 

### Anything else

_No response_

### Version

3.0.0-beta-1

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

here is the log:

[LOG-PATH]: /data/apache-dolphinscheduler-3.0.0-beta-1-bin/worker-server/logs/20220610/5803827748480_10-155-217.log, [HOST]:  Host{address='xxxx:1234', ip='xxxx', port=1234}
[INFO] 2022-06-10 08:46:55.800 +0800 [taskAppId=TASK-20220610-5803827748480_10-155-217] TaskLogLogger-class org.apache.dolphinscheduler.plugin.task.sql.SqlTask:[120] - Full sql parameters: SqlParameters{type='MYSQL', datasource=1, sql='select name from test.userinfo1 where name in (${name})', sqlType=0, sendEmail=null, displayRows=10, limit=0, segmentSeparator=, udfs='null', showType='null', connParams='null', groupId='0', title='null', preStatements=[], postStatements=[]}
[INFO] 2022-06-10 08:46:55.801 +0800 [taskAppId=TASK-20220610-5803827748480_10-155-217] TaskLogLogger-class org.apache.dolphinscheduler.plugin.task.sql.SqlTask:[121] - sql type : MYSQL, datasource : 1, sql : select name from test.userinfo1 where name in (${name}) , localParams : [],udfs : null,showType : null,connParams : null,varPool : [Property{prop='name', direct=IN, type=LIST, value='["lisi","zhangsan"]'}] ,query max result limit  0
[INFO] 2022-06-10 08:46:55.802 +0800 [taskAppId=TASK-20220610-5803827748480_10-155-217] TaskLogLogger-class org.apache.dolphinscheduler.plugin.task.sql.SqlTask:[91] - setSqlParamsMap: Property with paramName: name put in sqlParamsMap of content 
select name from test.userinfo1 where name in (${name}) successfully.
[INFO] 2022-06-10 08:46:55.802 +0800 [taskAppId=TASK-20220610-5803827748480_10-155-217] TaskLogLogger-class org.apache.dolphinscheduler.plugin.task.sql.SqlTask:[387] - after replace sql , preparing : 
select name from test.userinfo1 where name in (?)
[INFO] 2022-06-10 08:46:55.802 +0800 [taskAppId=TASK-20220610-5803827748480_10-155-217] TaskLogLogger-class org.apache.dolphinscheduler.plugin.task.sql.SqlTask:[396] - Sql Params are replaced sql , parameters:["lisi","zhangsan"](LIST)
[INFO] 2022-06-10 08:46:55.802 +0800 [taskAppId=TASK-20220610-5803827748480_10-155-217] TaskLogLogger-class org.apache.dolphinscheduler.plugin.task.sql.SqlTask:[465] - can't find udf function resource
[INFO] 2022-06-10 08:46:55.804 +0800 [taskAppId=TASK-20220610-5803827748480_10-155-217] TaskLogLogger-class org.apache.dolphinscheduler.plugin.task.sql.SqlTask:[369] - prepare statement replace sql : HikariProxyPreparedStatement@1139720980 wrapping com.mysql.cj.jdbc.ClientPreparedStatement: 
select name from test.userinfo1 where name in (** NOT SPECIFIED **) 
[INFO] 2022-06-10 08:46:55.805 +0800 [taskAppId=TASK-20220610-5803827748480_10-155-217] TaskLogLogger-class org.apache.dolphinscheduler.plugin.task.sql.SqlTask:[297] - main statement execute query, for sql: 
select name from test.userinfo1 where name in (?)
[ERROR] 2022-06-10 08:46:55.805 +0800 [taskAppId=TASK-20220610-5803827748480_10-155-217] TaskLogLogger-class org.apache.dolphinscheduler.plugin.task.sql.SqlTask:[213] - execute sql error: No value specified for parameter 1
[ERROR] 2022-06-10 08:46:55.805 +0800 [taskAppId=TASK-20220610-5803827748480_10-155-217] TaskLogLogger-class org.apache.dolphinscheduler.plugin.task.sql.SqlTask:[164] - sql task error: java.sql.SQLException: No value specified for parameter 1

--- a/dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/parser/ParameterUtils.java
+++ b/dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/parser/ParameterUtils.java
@@ -164,8 +164,12 @@ public class ParameterUtils {
         StringBuilder ret = new StringBuilder(split[0]);
         for (int i = 1; i < split.length; i++) {
             Property property = params.get(i);
+            String value = property.getValue();
             if (DataType.LIST.equals(property.getType())) {
-                List<Object> valueList = JSONUtils.toList(property.getValue(), Object.class);
+                List<Object> valueList = JSONUtils.toList(value, Object.class);
+                if (valueList.isEmpty() && StringUtils.isNotBlank(value)) {
+                    valueList.add(value);
+                }
                 valueList.forEach(v -> {
                     if (v instanceof String) {
                         ret.append("'").append(v).append("'");

--- a/dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/parser/ParameterUtils.java
+++ b/dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/parser/ParameterUtils.java
@@ -28,7 +28,6 @@ import org.apache.dolphinscheduler.spi.utils.JSONUtils;
 import org.apache.dolphinscheduler.spi.utils.StringUtils;
 
 import java.sql.PreparedStatement;
-import java.util.ArrayList;
 import java.util.Date;
 import java.util.HashMap;
 import java.util.Iterator;

--- a/dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/parser/ParameterUtils.java
+++ b/dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/parser/ParameterUtils.java
@@ -24,12 +24,15 @@ import static org.apache.dolphinscheduler.plugin.task.api.TaskConstants.PARAMETE
 import org.apache.dolphinscheduler.plugin.task.api.model.Property;
 import org.apache.dolphinscheduler.plugin.task.api.enums.DataType;
 import org.apache.dolphinscheduler.spi.utils.DateUtils;
+import org.apache.dolphinscheduler.spi.utils.JSONUtils;
 import org.apache.dolphinscheduler.spi.utils.StringUtils;
 
 import java.sql.PreparedStatement;
+import java.util.ArrayList;
 import java.util.Date;
 import java.util.HashMap;
 import java.util.Iterator;
+import java.util.List;
 import java.util.Map;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
@@ -146,6 +149,37 @@ public class ParameterUtils {
         } else if (dataType.equals(DataType.BOOLEAN)) {
             stmt.setBoolean(index, Boolean.parseBoolean(value));
         }
+    }
+
+    public static String replaceListParameter(Map<Integer, Property> params, String sql) {
+        if (params == null || params.isEmpty()) {
+            return sql;
+        }
+        List<Integer> removeList = new ArrayList<>();
+        String[] split = sql.split("\\?");
+        if (split.length == 0) {
+            return sql;
+        }
+        StringBuilder ret = new StringBuilder(split[0]);
+        for (int i = 1; i < split.length; i++) {
+            Property property = params.get(i);
+            if (DataType.LIST.equals(property.getType())) {
+                List<Object> valueList = JSONUtils.toList(property.getValue(), Object.class);
+                valueList.forEach(v -> {
+                    if (v instanceof String) {
+                        ret.append("'").append(v).append("'");
+                    } else {
+                        ret.append(v);
+                    }
+                    ret.append(",");
+                });
+                ret.deleteCharAt(ret.length() - 1);
+            } else {
+                ret.append("?");
+            }
+            ret.append(split[i]);
+        }
+        return ret.toString();
     }
 
     /**

--- a/dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/parser/ParameterUtils.java
+++ b/dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/parser/ParameterUtils.java
@@ -21,8 +21,8 @@ import static org.apache.dolphinscheduler.plugin.task.api.TaskConstants.PARAMETE
 import static org.apache.dolphinscheduler.plugin.task.api.TaskConstants.PARAMETER_FORMAT_TIME;
 import static org.apache.dolphinscheduler.plugin.task.api.TaskConstants.PARAMETER_SHECDULE_TIME;
 
-import org.apache.dolphinscheduler.plugin.task.api.model.Property;
 import org.apache.dolphinscheduler.plugin.task.api.enums.DataType;
+import org.apache.dolphinscheduler.plugin.task.api.model.Property;
 import org.apache.dolphinscheduler.spi.utils.DateUtils;
 import org.apache.dolphinscheduler.spi.utils.JSONUtils;
 import org.apache.dolphinscheduler.spi.utils.StringUtils;
@@ -152,7 +152,8 @@ public class ParameterUtils {
         }
     }
 
-    public static String replaceListParameter(Map<Integer, Property> params, String sql) {
+    public static String expandListParameter(Map<Integer, Property> params, String sql) {
+        Map<Integer, Property> expandMap = new HashMap<>();
         if (params == null || params.isEmpty()) {
             return sql;
         }
@@ -161,6 +162,7 @@ public class ParameterUtils {
             return sql;
         }
         StringBuilder ret = new StringBuilder(split[0]);
+        int index = 1;
         for (int i = 1; i < split.length; i++) {
             Property property = params.get(i);
             String value = property.getValue();
@@ -169,24 +171,42 @@ public class ParameterUtils {
                 if (valueList.isEmpty() && StringUtils.isNotBlank(value)) {
                     valueList.add(value);
                 }
-                valueList.forEach(v -> {
-                    if (v instanceof String) {
-                        ret.append("'").append(v).append("'");
-                    } else {
-                        ret.append(v);
+                for (int j = 0; j < valueList.size(); j++) {
+                    ret.append(PARAM_REPLACE_CHAR);
+                    if (j != valueList.size() - 1) {
+                        ret.append(",");
                     }
-                    ret.append(",");
-                });
-                ret.deleteCharAt(ret.length() - 1);
-                params.remove(i);
+                }
+                for (Object v : valueList ) {
+                    Property newProperty = new Property();
+                     if (v instanceof Integer) {
+                        newProperty.setType(DataType.INTEGER);
+                    } else if (v instanceof Long) {
+                        newProperty.setType(DataType.LONG);
+                    } else if (v instanceof Float) {
+                        newProperty.setType(DataType.FLOAT);
+                    } else if (v instanceof Double) {
+                        newProperty.setType(DataType.DOUBLE);
+                    } else {
+                        newProperty.setType(DataType.VARCHAR);
+                    }
+                    newProperty.setValue(v.toString());
+                    newProperty.setProp(property.getProp());
+                    newProperty.setDirect(property.getDirect());
+                    expandMap.put(index++, newProperty);
+                }
             } else {
                 ret.append(PARAM_REPLACE_CHAR);
+                expandMap.put(index++, property);
             }
             ret.append(split[i]);
         }
         if (PARAM_REPLACE_CHAR == sql.charAt(sql.length() - 1)) {
             ret.append(PARAM_REPLACE_CHAR);
+            expandMap.put(index, params.get(split.length));
         }
+        params.clear();
+        params.putAll(expandMap);
         return ret.toString();
     }
 

--- a/dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/parser/ParameterUtils.java
+++ b/dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/parser/ParameterUtils.java
@@ -51,6 +51,8 @@ public class ParameterUtils {
 
     private static final String DATE_START_PATTERN = "^[0-9]";
 
+    private static final char PARAM_REPLACE_CHAR = '?';
+
     private ParameterUtils() {
         throw new UnsupportedOperationException("Construct ParameterUtils");
     }
@@ -155,7 +157,6 @@ public class ParameterUtils {
         if (params == null || params.isEmpty()) {
             return sql;
         }
-        List<Integer> removeList = new ArrayList<>();
         String[] split = sql.split("\\?");
         if (split.length == 0) {
             return sql;
@@ -174,10 +175,14 @@ public class ParameterUtils {
                     ret.append(",");
                 });
                 ret.deleteCharAt(ret.length() - 1);
+                params.remove(i);
             } else {
-                ret.append("?");
+                ret.append(PARAM_REPLACE_CHAR);
             }
             ret.append(split[i]);
+        }
+        if (PARAM_REPLACE_CHAR == sql.charAt(sql.length() - 1)) {
+            ret.append(PARAM_REPLACE_CHAR);
         }
         return ret.toString();
     }

--- a/dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/test/java/org/apache/dolphinscheduler/plugin/task/api/parser/ParameterUtilsTest.java
+++ b/dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/test/java/org/apache/dolphinscheduler/plugin/task/api/parser/ParameterUtilsTest.java
@@ -1,3 +1,20 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
 package org.apache.dolphinscheduler.plugin.task.api.parser;
 
 import org.apache.dolphinscheduler.plugin.task.api.enums.DataType;

--- a/dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/test/java/org/apache/dolphinscheduler/plugin/task/api/parser/ParameterUtilsTest.java
+++ b/dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/test/java/org/apache/dolphinscheduler/plugin/task/api/parser/ParameterUtilsTest.java
@@ -28,10 +28,10 @@ public class ParameterUtilsTest {
         Assert.assertEquals("select * from test where col1 in ('c1','c2','c3') and date=? and col2 in ('d1','d2','d3')", sql);
 
         Map<Integer, Property> params2 = new HashMap<>();
-        params2.put(1, new Property(null, null, DataType.LIST, JSONUtils.toJsonString(Lists.newArrayList("c1", "c2", "c3"))));
+        params2.put(1, new Property(null, null, DataType.LIST, JSONUtils.toJsonString(Lists.newArrayList("c1"))));
         params2.put(2, new Property(null, null, DataType.DATE, "2020-06-30"));
         String sql2 = ParameterUtils.replaceListParameter(params2, "select * from test where col1 in (?) and date=?");
-        Assert.assertEquals("select * from test where col1 in ('c1','c2','c3') and date=?", sql2);
+        Assert.assertEquals("select * from test where col1 in ('c1') and date=?", sql2);
 
     }
 }

--- a/dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/test/java/org/apache/dolphinscheduler/plugin/task/api/parser/ParameterUtilsTest.java
+++ b/dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/test/java/org/apache/dolphinscheduler/plugin/task/api/parser/ParameterUtilsTest.java
@@ -32,19 +32,21 @@ import com.google.common.collect.Lists;
 public class ParameterUtilsTest {
 
     @Test
-    public void replaceListParameter() {
+    public void expandListParameter() {
         Map<Integer, Property> params = new HashMap<>();
         params.put(1, new Property(null, null, DataType.LIST, JSONUtils.toJsonString(Lists.newArrayList("c1", "c2", "c3"))));
         params.put(2, new Property(null, null, DataType.DATE, "2020-06-30"));
-        params.put(3, new Property(null, null, DataType.LIST, JSONUtils.toJsonString(Lists.newArrayList("d1", "d2", "d3"))));
-        String sql = ParameterUtils.replaceListParameter(params, "select * from test where col1 in (?) and date=? and col2 in (?)");
-        Assert.assertEquals("select * from test where col1 in ('c1','c2','c3') and date=? and col2 in ('d1','d2','d3')", sql);
+        params.put(3, new Property(null, null, DataType.LIST, JSONUtils.toJsonString(Lists.newArrayList(3.1415, 2.44, 3.44))));
+        String sql = ParameterUtils.expandListParameter(params, "select * from test where col1 in (?) and date=? and col2 in (?)");
+        Assert.assertEquals("select * from test where col1 in (?,?,?) and date=? and col2 in (?,?,?)", sql);
+        Assert.assertEquals(7, params.size());
 
         Map<Integer, Property> params2 = new HashMap<>();
         params2.put(1, new Property(null, null, DataType.LIST, JSONUtils.toJsonString(Lists.newArrayList("c1"))));
         params2.put(2, new Property(null, null, DataType.DATE, "2020-06-30"));
-        String sql2 = ParameterUtils.replaceListParameter(params2, "select * from test where col1 in (?) and date=?");
-        Assert.assertEquals("select * from test where col1 in ('c1') and date=?", sql2);
+        String sql2 = ParameterUtils.expandListParameter(params2, "select * from test where col1 in (?) and date=?");
+        Assert.assertEquals("select * from test where col1 in (?) and date=?", sql2);
+        Assert.assertEquals(2, params2.size());
 
     }
 }

--- a/dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/test/java/org/apache/dolphinscheduler/plugin/task/api/parser/ParameterUtilsTest.java
+++ b/dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/test/java/org/apache/dolphinscheduler/plugin/task/api/parser/ParameterUtilsTest.java
@@ -29,10 +29,6 @@ import org.junit.Test;
 
 import com.google.common.collect.Lists;
 
-/**
- * @author lirongqian
- * @date 2022/8/3
- **/
 public class ParameterUtilsTest {
 
     @Test

--- a/dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/test/java/org/apache/dolphinscheduler/plugin/task/api/parser/ParameterUtilsTest.java
+++ b/dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/test/java/org/apache/dolphinscheduler/plugin/task/api/parser/ParameterUtilsTest.java
@@ -1,0 +1,29 @@
+package org.apache.dolphinscheduler.plugin.task.api.parser;
+
+import org.apache.dolphinscheduler.plugin.task.api.enums.DataType;
+import org.apache.dolphinscheduler.plugin.task.api.model.Property;
+import org.apache.dolphinscheduler.spi.utils.JSONUtils;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import org.junit.Test;
+
+import com.google.common.collect.Lists;
+
+/**
+ * @author lirongqian
+ * @date 2022/8/3
+ **/
+public class ParameterUtilsTest {
+
+    @Test
+    public void replaceListParameter() {
+        Map<Integer, Property> params = new HashMap<>();
+        params.put(1, new Property(null, null, DataType.LIST, JSONUtils.toJsonString(Lists.newArrayList("c1", "c2", "c3"))));
+        params.put(2, new Property(null, null, DataType.DATE, "2020-06-30"));
+        params.put(3, new Property(null, null, DataType.LIST, JSONUtils.toJsonString(Lists.newArrayList("d1", "d2", "d3"))));
+        String sql = ParameterUtils.replaceListParameter(params, "select * from test where col1 in (?) and date=? and col2 in (?)");
+        System.out.println(sql);
+    }
+}

--- a/dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/test/java/org/apache/dolphinscheduler/plugin/task/api/parser/ParameterUtilsTest.java
+++ b/dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/test/java/org/apache/dolphinscheduler/plugin/task/api/parser/ParameterUtilsTest.java
@@ -7,6 +7,7 @@ import org.apache.dolphinscheduler.spi.utils.JSONUtils;
 import java.util.HashMap;
 import java.util.Map;
 
+import org.junit.Assert;
 import org.junit.Test;
 
 import com.google.common.collect.Lists;
@@ -24,6 +25,13 @@ public class ParameterUtilsTest {
         params.put(2, new Property(null, null, DataType.DATE, "2020-06-30"));
         params.put(3, new Property(null, null, DataType.LIST, JSONUtils.toJsonString(Lists.newArrayList("d1", "d2", "d3"))));
         String sql = ParameterUtils.replaceListParameter(params, "select * from test where col1 in (?) and date=? and col2 in (?)");
-        System.out.println(sql);
+        Assert.assertEquals("select * from test where col1 in ('c1','c2','c3') and date=? and col2 in ('d1','d2','d3')", sql);
+
+        Map<Integer, Property> params2 = new HashMap<>();
+        params2.put(1, new Property(null, null, DataType.LIST, JSONUtils.toJsonString(Lists.newArrayList("c1", "c2", "c3"))));
+        params2.put(2, new Property(null, null, DataType.DATE, "2020-06-30"));
+        String sql2 = ParameterUtils.replaceListParameter(params2, "select * from test where col1 in (?) and date=?");
+        Assert.assertEquals("select * from test where col1 in ('c1','c2','c3') and date=?", sql2);
+
     }
 }

--- a/dolphinscheduler-task-plugin/dolphinscheduler-task-sql/src/main/java/org/apache/dolphinscheduler/plugin/task/sql/SqlTask.java
+++ b/dolphinscheduler-task-plugin/dolphinscheduler-task-sql/src/main/java/org/apache/dolphinscheduler/plugin/task/sql/SqlTask.java
@@ -354,17 +354,15 @@ public class SqlTask extends AbstractTaskExecutor {
         boolean timeoutFlag = taskExecutionContext.getTaskTimeoutStrategy() == TaskTimeoutStrategy.FAILED
                 || taskExecutionContext.getTaskTimeoutStrategy() == TaskTimeoutStrategy.WARNFAILED;
         try {
-            Map<Integer, Property> params = sqlBinds.getParamsMap();
             PreparedStatement stmt = connection.prepareStatement(sqlBinds.getSql());
             if (timeoutFlag) {
                 stmt.setQueryTimeout(taskExecutionContext.getTaskTimeout());
             }
+            Map<Integer, Property> params = sqlBinds.getParamsMap();
             if (params != null) {
-                int index = 1;
                 for (Map.Entry<Integer, Property> entry : params.entrySet()) {
                     Property prop = entry.getValue();
-                    ParameterUtils.setInParameter(index, stmt, prop.getType(), prop.getValue());
-                    index++;
+                    ParameterUtils.setInParameter(entry.getKey(), stmt, prop.getType(), prop.getValue());
                 }
             }
             logger.info("prepare statement replace sql : {} ", stmt);

--- a/dolphinscheduler-task-plugin/dolphinscheduler-task-sql/src/main/java/org/apache/dolphinscheduler/plugin/task/sql/SqlTask.java
+++ b/dolphinscheduler-task-plugin/dolphinscheduler-task-sql/src/main/java/org/apache/dolphinscheduler/plugin/task/sql/SqlTask.java
@@ -355,7 +355,7 @@ public class SqlTask extends AbstractTaskExecutor {
                 || taskExecutionContext.getTaskTimeoutStrategy() == TaskTimeoutStrategy.WARNFAILED;
         try {
             Map<Integer, Property> params = sqlBinds.getParamsMap();
-            PreparedStatement stmt = connection.prepareStatement(ParameterUtils.replaceListParameter(params, sqlBinds.getSql()));
+            PreparedStatement stmt = connection.prepareStatement(sqlBinds.getSql());
             if (timeoutFlag) {
                 stmt.setQueryTimeout(taskExecutionContext.getTaskTimeout());
             }
@@ -431,9 +431,10 @@ public class SqlTask extends AbstractTaskExecutor {
         sql = replaceOriginalValue(sql, rgexo, paramsMap);
         // replace the ${} of the SQL statement with the Placeholder
         String formatSql = sql.replaceAll(rgex, "?");
+        // Convert the list parameter
+        formatSql = ParameterUtils.expandListParameter(sqlParamsMap, formatSql);
         sqlBuilder.append(formatSql);
-
-        // print repalce sql
+        // print replace sql
         printReplacedSql(sql, formatSql, rgex, sqlParamsMap);
         return new SqlBinds(sqlBuilder.toString(), sqlParamsMap);
     }

--- a/dolphinscheduler-task-plugin/dolphinscheduler-task-sql/src/main/java/org/apache/dolphinscheduler/plugin/task/sql/SqlTask.java
+++ b/dolphinscheduler-task-plugin/dolphinscheduler-task-sql/src/main/java/org/apache/dolphinscheduler/plugin/task/sql/SqlTask.java
@@ -25,7 +25,6 @@ import org.apache.dolphinscheduler.plugin.task.api.SQLTaskExecutionContext;
 import org.apache.dolphinscheduler.plugin.task.api.TaskConstants;
 import org.apache.dolphinscheduler.plugin.task.api.TaskException;
 import org.apache.dolphinscheduler.plugin.task.api.TaskExecutionContext;
-import org.apache.dolphinscheduler.plugin.task.api.enums.DataType;
 import org.apache.dolphinscheduler.plugin.task.api.enums.Direct;
 import org.apache.dolphinscheduler.plugin.task.api.enums.SqlType;
 import org.apache.dolphinscheduler.plugin.task.api.enums.TaskTimeoutStrategy;

--- a/dolphinscheduler-task-plugin/dolphinscheduler-task-sql/src/main/java/org/apache/dolphinscheduler/plugin/task/sql/SqlTask.java
+++ b/dolphinscheduler-task-plugin/dolphinscheduler-task-sql/src/main/java/org/apache/dolphinscheduler/plugin/task/sql/SqlTask.java
@@ -25,6 +25,7 @@ import org.apache.dolphinscheduler.plugin.task.api.SQLTaskExecutionContext;
 import org.apache.dolphinscheduler.plugin.task.api.TaskConstants;
 import org.apache.dolphinscheduler.plugin.task.api.TaskException;
 import org.apache.dolphinscheduler.plugin.task.api.TaskExecutionContext;
+import org.apache.dolphinscheduler.plugin.task.api.enums.DataType;
 import org.apache.dolphinscheduler.plugin.task.api.enums.Direct;
 import org.apache.dolphinscheduler.plugin.task.api.enums.SqlType;
 import org.apache.dolphinscheduler.plugin.task.api.enums.TaskTimeoutStrategy;
@@ -354,15 +355,17 @@ public class SqlTask extends AbstractTaskExecutor {
         boolean timeoutFlag = taskExecutionContext.getTaskTimeoutStrategy() == TaskTimeoutStrategy.FAILED
                 || taskExecutionContext.getTaskTimeoutStrategy() == TaskTimeoutStrategy.WARNFAILED;
         try {
-            PreparedStatement stmt = connection.prepareStatement(sqlBinds.getSql());
+            Map<Integer, Property> params = sqlBinds.getParamsMap();
+            PreparedStatement stmt = connection.prepareStatement(ParameterUtils.replaceListParameter(params, sqlBinds.getSql()));
             if (timeoutFlag) {
                 stmt.setQueryTimeout(taskExecutionContext.getTaskTimeout());
             }
-            Map<Integer, Property> params = sqlBinds.getParamsMap();
             if (params != null) {
+                int index = 1;
                 for (Map.Entry<Integer, Property> entry : params.entrySet()) {
                     Property prop = entry.getValue();
-                    ParameterUtils.setInParameter(entry.getKey(), stmt, prop.getType(), prop.getValue());
+                    ParameterUtils.setInParameter(index, stmt, prop.getType(), prop.getValue());
+                    index++;
                 }
             }
             logger.info("prepare statement replace sql : {} ", stmt);
@@ -370,7 +373,6 @@ public class SqlTask extends AbstractTaskExecutor {
         } catch (Exception exception) {
             throw new TaskException("SQL task prepareStatementAndBind error", exception);
         }
-
     }
 
     /**

